### PR TITLE
Fixes for release scripts

### PR DESCRIPTION
**Pull request by mikedougherty**
I had to make several changes in preparation for the 1.9-cs release. The commit messages should be fairly explanatory but I'm happy to provide more info if necessary.

@jfrazelle i see this might conflict with #17699, let me know how you'd like to proceed wrt that.

### CODE REVIEW

**Comment by jfrazelle** [comment]
No worries we can merge yours first and I'll rebase, I'll take a look tomorrow thanks!!






On Wed, Nov 4, 2015 at 5:51 PM -0800, "Michael Dougherty" <notifications@github.com> wrote:












I had to make several changes in preparation for the 1.9-cs release. The commit messages should be fairly explanatory but I'm happy to provide more info if necessary.



@jfrazelle i see this might conflict with #17699, let me know how you'd like to proceed wrt that.



You can view, comment on, or merge this pull request online at:


  https://github.com/docker/docker/pull/17714

Commit Summary

  Add missing env var to documentation
  Add .docker to gitignore
  Add old docker-engine-cs name to package conflicts
  Fix several problems in repo make targets
  Fix some issues with deb repository creation from scratch
  Add a flag to allow keeping bundles around (helps with CI)


File Changes

  
    M
    .gitignore
    (1)
  
  
    M
    contrib/reprepro/suites.sh
    (2)
  
  
    M
    hack/make.sh
    (2)
  
  
    M
    hack/make/.build-deb/control
    (2)
  
  
    M
    hack/make/.build-rpm/docker-engine.spec
    (1)
  
  
    M
    hack/make/release-deb
    (47)
  
  
    M
    hack/make/sign-repos
    (4)
  
  
    M
    hack/release.sh
    (1)
  


Patch Links:

  https://github.com/docker/docker/pull/17714.patch
  https://github.com/docker/docker/pull/17714.diff

**Comment by tiborvass** [comment]
Ping @tianon

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,4 @@ man/man5
 man/man8
 pyenv
 vendor/pkg/
+.docker
```
Comment (tianon) [inline .gitignore]: What's this about?

File: contrib/reprepro/suites.sh [hunk 2]
```diff
@@ -5,7 +5,7 @@ cd "$(dirname "$BASH_SOURCE")/../.."
 
 targets_from() {
        git fetch -q https://github.com/docker/docker.git "$1"
-       git ls-tree -r --name-only origin/master contrib/builder/deb | grep '/Dockerfile$' | sed -r 's!^contrib/builder/deb/|-debootstrap|/Dockerfile$!!g'
+       git ls-tree -r --name-only "$(git rev-parse FETCH_HEAD)" contrib/builder/deb | grep '/Dockerfile$' | sed -r 's!^contrib/builder/deb/|-debootstrap|/Dockerfile$!!g'
```
Comment (tianon) [inline contrib/reprepro/suites.sh]: Doh!  Great catch!

File: hack/make/.build-deb/control [hunk 3]
```diff
@@ -13,7 +13,7 @@ Recommends: aufs-tools,
             git,
             xz-utils,
             ${apparmor:Recommends}
-Conflicts: docker (<< 1.5~), docker.io, lxc-docker, lxc-docker-virtual-package
+Conflicts: docker (<< 1.5~), docker.io, lxc-docker, lxc-docker-virtual-package, docker-engine-cs
```
Comment (tianon) [inline hack/make/.build-deb/control]: IMO, this shouldn't be listed as a conflict in the open source package, but in the `docker-engine-cs` package.

File: hack/make/release-deb [hunk 4]
```diff
@@ -23,7 +23,29 @@ mkdir -p "$APTDIR/conf" "$APTDIR/db"
...
+done
+
+# set the component for the version being released
+component="main"
+
+if [[ "$VERSION" == *-rc* ]]; then
+	component="testing"
+fi
+
+if [ $DOCKER_EXPERIMENTAL ] || [[ "$VERSION" == *-dev ]] || [ -n "$(git status --porcelain)" ]; then
```
Comment (tianon) [inline hack/make/release-deb]: `$DOCKER_EXPERIMENTAL` needs quotes around it

File: hack/make/release-deb [hunk 5]
```diff
@@ -70,16 +92,14 @@ if [ ! -f "$APTDIR/conf/docker-engine-release.conf" ]; then
...
-	component="experimental"
-fi
+for dir in contrib/builder/deb/*/; do
+	version="$(basename "$dir")"
+	codename="${version//debootstrap-}"
+	mkdir -p "$APTDIR/pool/$component/d/docker-engine/"
+	for arch in "${arches[@]}"; do
+		mkdir -p "$APTDIR/dists/$codename/$component/binary-$arch"
+	done
+done
```
Comment (tianon) [inline hack/make/release-deb]: Is there any particular reason we must pre-create all of these and not just create them with `mkdir -p` on-demand in the loops below?

File: hack/make/sign-repos [hunk 6]
```diff
@@ -26,7 +26,7 @@ sign_packages(){
 
 		# sign the repo metadata
 		for F in $(find $APTDIR -name Release); do
-			if test "$F" -nt "$F.gpg" ; then
+			if ! test -e "$F.gpg" || test "$F" -nt "$F.gpg" ; then
```
Comment (tianon) [inline hack/make/sign-repos]: Doesn't `-nt` already "do the right thing" here?  If `$F.gpg` doesn't exist, `$F` should be "newer":
```console
$ ls
$ touch a
$ test a -nt b && echo yes || echo no
yes
$ test b -nt a && echo yes || echo no
no
```

File: hack/make/.build-deb/control (hunk 3 above)
Comment (mikedougherty) [inline hack/make/.build-deb/control]: A decision was made to not actually have different package names, but instead just let the version string speak for itself. So depending on the repository you have added to your package manager, you would get either:

```bash
$ docker --version
Docker version 1.9.0, build 76d6bc9
```

or
```bash
$ docker --version
Docker version 1.9.0-cs1-rc1, build e1417a4
```

Plus, they do actually conflict. You can't install the open-source version and the CS version at the same time. The second install will clobber `$(which docker)` the first one installed. And at this point, docker-engine-cs is basically an old package that users might have installed that does conflict with current things, just like the other names here (docker.io, lxc-docker).

File: hack/make/release-deb (hunk 5 above)
Comment (mikedougherty) [inline hack/make/release-deb]: TBH I was having all kinds of trouble going from an empty directory to a populated repo and this was just one of the things I did to try to fix it. I'm not 100% sure it's necessary but I didn't feel it was a terrible change so I left it in. If it's genuinely unwanted I can work on getting rid of this chunk (and related) from the diff.

File: hack/make/release-deb (hunk 4 above)
Comment (mikedougherty) [inline hack/make/release-deb]: this block is just moved from L80, but sure ;)

File: hack/make/sign-repos (hunk 6 above)
Comment (mikedougherty) [inline hack/make/sign-repos]: This one also falls under "all kinds of trouble" mentioned earlier. Also apparently zsh and bash behave differently in this regard, which is very frustrating to learn:

```bash
$ echo $SHELL
/bin/zsh
$ ls
$ touch a
$ test a -nt b && echo yes || echo no
no
$ test b -nt a && echo yes || echo no
no
```

You are right though, from https://www.mkssoftware.com/docs/man1/test.1.asp:
> Note:
All existing files are considered to be newer than a nonexistent file. In turn, no nonexistent file is newer than any existing file. Two nonexistent files are considered the same age.

I'll revert this change (and the yum one below) and run some tests to verify that I'm not crazy :)

File: .gitignore (hunk 1 above)
Comment (mikedougherty) [inline .gitignore]: from https://github.com/mikedougherty/docker/commit/63034961a930e8876caac96c107690a689367ae1

> Jenkins uses the workspace for HOME/.docker (for parallel builds). Without ignoring this directory no built version is ever 'clean'

File: hack/make/release-deb (hunk 5 above)
Comment (mikedougherty) [inline hack/make/release-deb]: FYI I went ahead and got rid of this change. I sprinkled a couple `mkdir`s around to compensate.

**Comment by mikedougherty** [comment]
Made some changes; rebased and updated.

File: hack/make.sh [hunk 7]
```diff
@@ -287,7 +287,7 @@ main() {
 	# We want this to fail if the bundles already exist and cannot be removed.
 	# This is to avoid mixing bundles from different versions of the code.
 	mkdir -p bundles
-	if [ -e "bundles/$VERSION" ]; then
+	if [ -e "bundles/$VERSION" ] && [ -z "$KEEPBUNDLE" ]; then
```
Comment (jfrazelle) [inline hack/make.sh]: ah nice

File: hack/make/release-deb [hunk 8]
```diff
@@ -53,7 +75,7 @@ if [ ! -f "$APTDIR/conf/apt-ftparchive.conf" ]; then
 	for suite in $(exec contrib/reprepro/suites.sh); do
 		cat <<-EOF
 		Tree "dists/${suite}" {
-			Sections "main testing experimental";
+			Sections "${components[*]}";
```
Comment (jfrazelle) [inline hack/make/release-deb]: nice

**Comment by jfrazelle** [comment]
LGTM

**Comment by mikedougherty** [comment]
Fixed quotes around DOCKER_EXPERIMENTAL for real and force pushed because I am bad at rebasing.

**Comment by tiborvass** [comment]
Ping @tianon

File: .gitignore (hunk 1 above)
Comment (tianon) [inline .gitignore]: Hmm, so why not create a global `gitignore` or just `.git/info/exclude` in the cloned repo? (https://help.github.com/articles/ignoring-files/#create-a-global-gitignore)

I feel like this is a pretty strong edge case that most people won't be hitting, so it feels strange to put something so environmental in the repo-managed `.gitignore`.

Comment (LK4D4) [inline .gitignore]: I agree, it's weird.

**Comment by mikedougherty** [comment]
Thanks for the feedback guys, I'm travelling a bit but will update this and ping again in the next couple days

**Comment by mikedougherty** [comment]
Removed the .gitignore commit.

@tianon have you had a chance to consider my response about the package conflicts change?

**Comment by calavera** [comment]
LGTM.

Can you squash the commits so we can easily isolate this changeset?

**Comment by mikedougherty** [comment]
Done and done :)

**Comment by jfrazelle** [comment]
LGTM

**Comment by tianon** [comment]
LGTM; thanks :+1:

**Comment by thaJeztah** [comment]
Thanks @mikedougherty !

**Pull request merged**